src/lib/work_partitioner/job_pool.ml

```diff
@@ -1,42 +1,60 @@
 open Core_kernel
 
-type ('accum, 'final) fold_action =
-  | Continue of 'accum
-  | Continue_remove of 'accum
-  | Stop of 'final
-  | Stop_remove of 'final
-
-module Make (Id : Map.Key) (Spec : T) = struct
+module Make (Id : Hashtbl.Key) (Spec : T) = struct
   type job = (Spec.t, Id.t) Snark_work_lib.With_job_meta.t
 
-  module IdMap = Map.Make (Id)
-
-  type t = job IdMap.t
-
-  let first_job t : job option = IdMap.min_elt t |> Option.map ~f:Tuple2.get2
-
-  let fold_until ~init ~f ~finish t =
-    let seq = IdMap.to_sequence ~order:`Increasing_key t in
-    let handle_item (acc, retained) (k, v) =
-      match f acc v with
-      | Continue acc_new ->
-          Continue_or_stop.Continue (acc_new, retained)
-      | Continue_remove acc_new ->
-          Continue (acc_new, IdMap.remove retained k)
-      | Stop final ->
-          Stop (final, retained)
-      | Stop_remove final ->
-          Stop (final, IdMap.remove retained k)
+  type t =
+    { timeline : Id.t Deque.t (* For iteration *)
+    ; index : (Id.t, job) Hashtbl.t (* For marking job as done *)
+    }
+
+  let create () =
+    { timeline = Deque.create (); index = Hashtbl.create (module Id) }
+
+  let remove ~id t = Hashtbl.find_and_remove t.index id
+
+  let find ~id t = Hashtbl.find t.index id
+
+  let change_inplace ~id ~f t = Hashtbl.change t.index id ~f
+
+  let set_inplace ~id ~job t = Hashtbl.set t.index ~key:id ~data:job
```

Comment:
Okay, now we use `set`, which may add a new element to the index, while the item won't be present in the timeline.

This seems to violate the invariant of the structure (as I understand it), that any element from, index is also present in the timeline.

BTW, let's describe this invariant in a comment for `type t`.

Reply:
If we need `set_inplace` that only modifies an id that exists, let's implement it as `change_inplace ~id ~f:(const (Some job))`. And update the doc-comment.

If the current doc-comment is what we want, let's update the implementation to ensure the invariant is not violated.